cheroot/connections.py

```diff
@@ -0,0 +1,254 @@
+
+import io
+import select
+import six
+import socket
+import time
+
+from . import errors
+from .makefile import MakeFile
+
+from collections import OrderedDict
+
+
+try:
+    import fcntl
+except ImportError:
+    try:
+        from ctypes import windll, WinError
+        import ctypes.wintypes
+        _SetHandleInformation = windll.kernel32.SetHandleInformation
+        _SetHandleInformation.argtypes = [
+            ctypes.wintypes.HANDLE,
+            ctypes.wintypes.DWORD,
+            ctypes.wintypes.DWORD,
+        ]
+        _SetHandleInformation.restype = ctypes.wintypes.BOOL
+    except ImportError:
+        def prevent_socket_inheritance(sock):
+            """Stub inheritance prevention.
+
+            Dummy function, since neither fcntl nor ctypes are available.
+            """
+            pass
+    else:
+        def prevent_socket_inheritance(sock):
+            """Mark the given socket fd as non-inheritable (Windows)."""
+            if not _SetHandleInformation(sock.fileno(), 1, 0):
+                raise WinError()
+else:
+    def prevent_socket_inheritance(sock):
+        """Mark the given socket fd as non-inheritable (POSIX)."""
+        fd = sock.fileno()
+        old_flags = fcntl.fcntl(fd, fcntl.F_GETFD)
+        fcntl.fcntl(fd, fcntl.F_SETFD, old_flags | fcntl.FD_CLOEXEC)
+
+
+class ConnectionManager:
+    """Class which manages HTTPConnection objects.
+
+    This is for connections which are being kept-alive for follow-up requests.
+    """
+
+    def __init__(self, server):
+        """Initialize ConnectionManager object.
+
+        Args:
+            server (cheroot.server.HTTPServer): web server object
+                that uses this ConnectionManager instance.
+        """
+        self.server = server
+        self.connections = OrderedDict()
+
+    def put(self, conn):
+        """Put idle connection into the ConnectionManager to be managed.
+
+        Args:
+            conn (cheroot.server.HTTPConnection): HTTP connection
+                to be managed.
+        """
+        self.connections[conn] = time.time(), conn.rfile.has_data()
+
+    def expire(self):
+        """Expire least recently used connections.
+
+        This happens if there are either too many open connections, or if the
+        connections have been timed out.
+
+        This should be called periodically.
+        """
+        if not self.connections:
+            return
+
+        # Check for too many open connections.
+        conns = list(self.connections.items())
+        ka_limit = self.server.keep_alive_conn_limit
+        if ka_limit is not None:
+            [self._close(conn[0]) for conn in conns[:-ka_limit]]
+            conns = conns[-ka_limit:]
+
+        # Check for old connections.
+        now = time.time()
+        for (conn, (ctime, _)) in conns:
+            # Oldest connection out of the currently available ones.
+            if (ctime + self.server.timeout) < now:
+                self._close(conn)
```

Comment:
You close connections but don't delete them from `self.connections`, do you? Wouldn't this cause a memory leak?

Reply:
Oh, you just remove them inside of that call...